### ISSUE TITLE
Special characters render password unuseable.

### ISSUE THREAD

**Issue by cobrafast** [nZEDb/nZEDb#2249]
Passwords containing special characters simply do not work as expected.

Very likely caused by the use of `htmlspecialchars()` here https://github.com/nZEDb/nZEDb/blob/0.x/www/pages/register.php#L38

**Comment by b3rs3rk** [comment]
http://stackoverflow.com/questions/7792038/how-to-properly-escape-password-field-on-html-form

These folks are under the impression that htmlspecialchars() doesn't affect the password posting.

**Comment by cobrafast** [comment]
Well yes, to an HTML interpreter `>` and `&gt;` are the same, but to MySQL or some hash algorithm they're not.
If you look at the code I linked, the converted password gets sent to the user class to create a new user (`$ret = $page->users->signUp(...$password...`), however the login page tries to compare the passwords without the `htmlspecialchars()` conversion.

Here's an example to clarify:
1. Sign up with password `Hi123&`
2. Password gets converted to `Hi123&amp;`, then hashed and saved to db
3. Log in with password `Hi123&`
4. Password gets hashed and compared
5. Login fails

* https://github.com/nZEDb/nZEDb/blob/0.x/www/pages/register.php#L52
* https://github.com/nZEDb/nZEDb/blob/0.x/nzedb/Users.php#L232
* https://github.com/nZEDb/nZEDb/blob/0.x/www/pages/login.php#L27
* https://github.com/nZEDb/nZEDb/blob/0.x/nzedb/Users.php#L778

**Comment by b3rs3rk** [comment]
Well, the easy answer is to make sure all authentication attempts utilize htmlspecialchars().  The more difficult one is to change the way the password is received/hashed/stored because that will require a full reset of all user passwords.  I'm unable to replicate your findings though.  I created a user with password `Hi!@#456&*(` and can login with said user just fine.

**Comment by b3rs3rk** [comment]
Doing a var_dump of the password after it is checked on login doesn't show any URL encoding with that password.  And I'm var_dumping the exact argument contents of what is sent to the function that compares the password to the hash stored in the DB.

I should add that I did the var_dump inside the gate that denotes the check succeeded and the user is to be logged in.

**Comment by cobrafast** [comment]
That's exactly my point.
If you register through the register page the password _is_ converted, if you log in through the login page the password _is not_ converted, so the hashes shouldn't match.

**Comment by b3rs3rk** [comment]
I follow now.  Double-checked the admin site code and we're not doing htmlspecialchars() there when we're creating users.  I'd suggest a patch against dev with a PR to remove that function from the registration page.

**Comment by b3rs3rk** [comment]
Hmm, it looks like dee9ec21c5c80d8bf52c647380ed0b63027e7085 changed this to protect from XSS.

... But that doesn't make sense in the least.  XSS prevention isn't done on input.  It is done on output.  As long as smarty is escaping the output when it is displaying the information to the client browser, it is effectively protecting against XSS.

**Comment by DariusIII** [comment]
An input from reddit user:

> I saw that but it's the incorrect way to do it. The correct way would be to not output the password back on the field and make the user re-type it in. If you want to prevent that then use client side validation.
> You can also JUST escape on output. Problem you have is you're escaping too early.

**Comment by b3rs3rk** [comment]
Yeah, what they are saying is what I am saying.  We should be ensuring all form entries in the Smarty templates are escaping their set values on display to the client.  Do it all in one place, not in PHP and Smarty.  Always in Smarty or Always in PHP.

**Comment by b3rs3rk** [comment]
Ah, here is the answer.  http://www.smarty.net/docs/en/variable.escape.html.tpl

If we set this variable within Smarty it will automatically escape all variable assignment outputs.  We could do this for all forms and be XSS safe.

**Comment by cobrafast** [comment]
Alternatively the smarty templates could be reworked so variable placements read like `{$var|escape}`

* http://www.smarty.net/docsv2/en/language.modifier.escape

**Comment by b3rs3rk** [comment]
We do that in many places already.  But with a template that includes form data, we should be escaping the entire template globally instead of wasting time escaping each one.

**Comment by b3rs3rk** [comment]
After trying it out, we have to individually escape them.  Setting it globally meeses with some of the theme's formatting.
